### PR TITLE
Shared Semantic Model

### PR DESCRIPTION
**Pull request by bauer-andreas**
This PR adds a first prototype for the shared semantic model (including a REST exporter working on that model). This is still a draft and not meant to be merged. It should be a place to follow progress and start discussions about the different needs of all the different exporters. ~The branch is based on #17 which is yet to be merged.~

In theory all interface exporters should already be able to work with this shared semantic model, as there are still references maintained to the `ContextNode` of the every registered `Component` (apart from the fact, that most of the Apodini internals still heavily depend on Vapor). Any dependence on the `Context` (and on Vapor) should be removed over time and all requirements should be built right into the model itself.

EDIT (2020-12-01): #17 is now merged.

### CODE REVIEW

**Comment by bauer-andreas** [comment]
The shared model is now at a stage where we think it would make sense to get a first version merged.
A first version of the #23 proposal was implemented by @lschlesinger.
`Guards`, `Handlers` (yet to replace handling `Component`s) and `ResponseTransformers` are not yet modeled as #32  and #26 are set to be finished/implemented.
The shared model already generates some relationship information as described in #41 but additional ways mentioned in the proposal are set to be implemented with a separate PR.

Our goal would be to get this version of the shared model merged (hopefully sometime this week) and start adopting it in other exporters, OpenAPI probably being first, but in general anyone could adopt that. This will give us a better insight into the requirements needed by the exporters and allows us to incorporate those step by step.

So would be great if you could have a look at it @PSchmiedmayer @nerdsupremacist or whoever wants to throw in their opinion.

@nerdsupremacist Could you maybe specifically have a look at that following two code snippets. @lschlesinger and me struggled on how to properly check for the `Parameter` type (as the generic `Element` is unknown at that point). Maybe you have a solution on how to better tackle that problem.
https://github.com/Apodini/Apodini/blob/2c7649ed5cc1c59001c71f396c7b1f1862d180eb/Sources/Apodini/Semantic%20Model%20Builder/EndpointParameter.swift#L34-L46
https://github.com/Apodini/Apodini/blob/2c7649ed5cc1c59001c71f396c7b1f1862d180eb/Sources/Apodini/Semantic%20Model%20Builder/EndpointsTree.swift#L98-L107

File: Sources/Apodini/Components/PathComponent.swift [hunk 1]
```diff
@@ -39,29 +41,32 @@ struct StringPathBuilder: PathBuilder {
...
+        self
+    }
+
     func append<P>(to pathBuilder: inout P) where P: PathBuilder {
         pathBuilder.append(self)
     }
 }
 
 /// Used to define parameter in a path
 public struct Identifier<T: Identifiable> {
```
Comment (PSchmiedmayer) [inline Sources/Apodini/Components/PathComponent.swift]: I think that the concept of an `Identifier` would be no longer needed as we have the `@PathParameter` and therefore I would propose to remove it here.

File: Sources/Apodini/Request/RequestInjectable.swift [hunk 2]
```diff
@@ -11,25 +11,39 @@ import Runtime
...
-    mutating func inject(using request: Vapor.Request, with decoder: SemanticModelBuilder?) throws
+    mutating func inject(using request: Vapor.Request, with decoder: RequestInjectableDecoder?) throws
+}
+
+protocol RequestInjectableDecoder {
+    func decode<T: Decodable>(_ type: T.Type, from request: Vapor.Request) throws -> T?
+}
+
+// swiftlint:disable:next todo
+// TODO Is there ANY better location to place this than on global?
```
Comment (PSchmiedmayer) [inline Sources/Apodini/Request/RequestInjectable.swift]: "ANY better location" 😅 ... I think an alternative would be to make the method private and add extensions for different types that need to support this functionality, e.g. `Component.extractRequestInjectables() -> [String: RequestInjectable]`.
If this isn't possible a global scope is fine for me as long as it is not public.

File: Sources/Apodini/Semantic Model Builder/EndpointsTree.swift [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            endpoint.treeNode = self
+            endpoints[endpoint.operation] = endpoint
+        } else {
+            var pathComponents = paths
+            if let first = pathComponents.removeFirst() as? _PathComponent {
+                var child = nodeChildren[first.description]
+                if child == nil {
+                    // swiftlint:disable:next todo
+                    // TODO is there a better way to check for type Parameter instead of string comparison?
+                    if let info = try? typeInfo(of: type(of: first)), info.mangledName == "Parameter" {
```
Comment (PSchmiedmayer) [inline Sources/Apodini/Semantic Model Builder/EndpointsTree.swift]: Maybe you could use something like `let parameterType = try? typeInfo(of: Parameter<Int>.self)` to get the type information of a Parameter type. This would throw a compile error if the `Parameter` type would change.

File: Sources/Apodini/Semantic Model Builder/EndpointsTree.swift [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+                        let mirror = Mirror(reflecting: first)
+
+                        // swiftlint:disable:next force_cast
+                        let options = mirror.children.first { $0.label == "options" }!.value as! PropertyOptionSet<ParameterOptionNameSpace>
```
Comment (PSchmiedmayer) [inline Sources/Apodini/Semantic Model Builder/EndpointsTree.swift]: I think we should add some enum with static properties that stores these constants such as `"options"` and `"Parameter"` where they are defined/in the `Parameter` type and use it here. Then we would at least have one file that stores all this context.

File: Sources/Apodini/Semantic Model Builder/EndpointsTree.swift (hunk 3 above)
Comment (PSchmiedmayer) [reply Sources/Apodini/Semantic Model Builder/EndpointsTree.swift]: The same would apply to this line here: https://github.com/Apodini/Apodini/blob/2c7649ed5cc1c59001c71f396c7b1f1862d180eb/Sources/Apodini/Semantic%20Model%20Builder/EndpointParameter.swift#L38

File: Tests/ApodiniTests/SharedSemanticModelBuilderTests.swift [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            "Hello \(name)"
+        }
+    }
+    
+    struct TestComponent: Component {
+        @PathParameter
+        var name: String
+        
+        var content: some Component {
+            Group("a") {
```
Comment (PSchmiedmayer) [inline Tests/ApodiniTests/SharedSemanticModelBuilderTests.swift:62]: I would also suggest to add a test case where there is a `Group` and some other component in the `Group("a")` just to make sure we also cover this edge case.

**Review by PSchmiedmayer — CHANGES_REQUESTED** [review]
@Supereg Thanks for the PR and the new shared Semantic Model! I like the approach and only had a few small comments and suggestions in a few places. Apart from these small comments I think that the code is ready to be merged 🚀

File: Sources/Apodini/Request/RequestInjectable.swift (hunk 2 above)
Comment (bauer-andreas) [reply Sources/Apodini/Request/RequestInjectable.swift]: Solved with 32d549e

File: Sources/Apodini/Components/PathComponent.swift (hunk 1 above)
Comment (bauer-andreas) [reply Sources/Apodini/Components/PathComponent.swift]: Solved via 263ffed
The PathBuilder now also properly supports `@Parameter` as it conforms to `_PathComponent` as well.

File: Sources/Apodini/Semantic Model Builder/EndpointsTree.swift (hunk 3 above)
Comment (bauer-andreas) [reply Sources/Apodini/Semantic Model Builder/EndpointsTree.swift]: Do I understand you correct, the idea is to add a line like you suggested which directly references `Parameter`, but **isn't used** actually, but serves as a hint when Parameter would change? Or what is the idea here?
Not sure if this solves our problem in the days of automatic refactoring? Most editors (like Xcode) probably support renaming all **used** references, only few IDEs search for string usage as well.

File: Tests/ApodiniTests/SharedSemanticModelBuilderTests.swift (hunk 5 above)
Comment (bauer-andreas) [reply Tests/ApodiniTests/SharedSemanticModelBuilderTests.swift:62]: I assume this is solved via a0f36a5 by @lschlesinger

File: Sources/Apodini/Semantic Model Builder/EndpointsTree.swift (hunk 3 above)
Comment (lschlesinger) [reply Sources/Apodini/Semantic Model Builder/EndpointsTree.swift]: As far as I've understood, we should do this to avoid the String Comparison:
```
let parameterType = try? typeInfo(of: Parameter<String>.self)
if let info = try? typeInfo(of: type(of: first)), info.mangledName == parameterType?.mangledName { 
/// ...do something
}
```

where String can be replaced by Int, doesn't matter.
I've tested it just now, it works and tests are still succeeding.

File: Tests/ApodiniTests/SharedSemanticModelBuilderTests.swift (hunk 5 above)
Comment (lschlesinger) [reply Tests/ApodiniTests/SharedSemanticModelBuilderTests.swift:62]: yes, just added this corner case!

File: Sources/Apodini/Semantic Model Builder/EndpointsTree.swift (hunk 4 above)
Comment (bauer-andreas) [reply Sources/Apodini/Semantic Model Builder/EndpointsTree.swift]: Could you maybe illustrate how you want this to look? Just an empty enum with `static let` definitions?

File: Sources/Apodini/Semantic Model Builder/EndpointsTree.swift (hunk 3 above)
Comment (bauer-andreas) [reply Sources/Apodini/Semantic Model Builder/EndpointsTree.swift]: ah yeah, my head went to a complete different direction 😂 
@lschlesinger do you want to make those changes to resolve this ☺️?

Comment (lschlesinger) [reply Sources/Apodini/Semantic Model Builder/EndpointsTree.swift]: solved with 298e00d53fa54ce3e7286d83549ac436a4a2d87b

File: Sources/Apodini/Semantic Model Builder/EndpointsTree.swift (hunk 4 above)
Comment (bauer-andreas) [reply Sources/Apodini/Semantic Model Builder/EndpointsTree.swift]: I hope 967e95c is what you had in mind?

**Review by PSchmiedmayer — APPROVED** [review]
